src/posts/2021-11-19-ecto_query_cheatsheet.md

```diff
@@ -0,0 +1,180 @@
+iex> Repo
+BikeBrigade.Repo
+```
+
+#### Queries
+
+```elixir
+alias BikeBrigade.Repo
+alias BikeBrigade.Accounts.User
+import Ecto.Query
+```
+
+```elixir
+# Find one by ID
+User |> Repo.get(1)
+
+# Find one by another attribute
+User |> Repo.get_by(name: "Dispatcher McGee")
+
+# Find first
+User
+|> first
+|> Repo.one
+
+# List all
+User |> Repo.all
+
+# WHERE (by attribute)
+User
+|> where(name: "Dispatcher McGee")
+|> Repo.all
+
+# WHERE (with another operator)
+{:ok, today} = DateTime.now("Etc/UTC")
+User
+|> where([u], u.updated_at < ^today)
+|> Repo.all
+
+# WHERE (with a LIKE operator)
+User
+|> where([u], ilike(u.name, "%McGee%"))
+|> Repo.all
+
+# COUNT
+User
+|> Repo.aggregate(:count, :id)
+
+# COUNT (by group)
+User
+|> group_by(:name)
+|> select([u], [u.name, count(u.id)])
+|> Repo.all
+
+# DISTINCT (get unique values)
+User
+|> distinct([u], u.name)
+|> Repo.all
+```
+
+#### Mutations
+
+```elixir
+# Create
+%User{name: "Stephanie"} |> Repo.insert
+```
+###
+The code above works, but bypasses any validations or constraints we've created. This is where [Ecto.Changeset](https://hexdocs.pm/ecto/Ecto.Changeset.html) comes in.
+
+```elixir
+import Ecto.Changeset
+```
+```elixir
+# Create
+User.changeset(%User{}, %{name: "Stephanie"})
+|> Repo.insert
+# => returns errors
+
+# Create (using all required fields)
+User.changeset(%User{}, %{name: "Stephanie", email: "stephanie@mail.com", phone: "+16479231256"})
+|> Repo.insert
+
+# Update
+stephanie = User |> Repo.get_by(name: "Stephanie")
+User.changeset(stephanie, %{name: "Stephanopolus"})
+|> Repo.update
+
+# Delete
+stephanie = User |> Repo.get_by(name: "Stephanie")
+stephanie |> Repo.delete
+
+# Multiple changes within a transaction
+mcgee = User |> Repo.get_by(name: "Dispatcher McGee")
+stephanie = User |> Repo.get_by(name: "Stephanie")
+
+Repo.transaction(fn ->
+  User.changeset(stephanie, %{email: "bikesrock@mail.com"}) |> Repo.update
+  User.changeset(mcgee, %{email: "bikesrock@mail.com"}) |> Repo.update
+end)
+
+# => returns errors because of uniqueness constraint. Neither records are updated.
+```
+
+#### Associations
+
+Referencing associations is only allowed when explicitly loaded. For example, the following code will result in an error:
+```elixir
+iex> alias BikeBrigade.Delivery.Campaign
+iex> camapign = Campaign |> Repo.get(1)
+iex> campaign.tasks
+#Ecto.Association.NotLoaded<association :tasks is not loaded>
+
+# To fix, we explicity query for the tasks along with the campaign
+iex> campaign = Campaign |> Repo.get(1) |> Repo.preload(:tasks)
+iex> campaign.tasks
+[...]
+```
+
+Persisting associations
+
+```elixir
+iex> alias BikeBrigade.Riders.{Rider, Tag}
+iex> rider = Rider |> Ecto.Query.first |> Repo.one |> Repo.preload(:tags)
+iex> rider.tags
+[]
+
+iex> rapid_response_tag = Repo.insert! %Tag{name: "rapid-responder"}
+iex> rider_changeset = Ecto.Changeset.change(rider)
+iex> rider_with_tags = Ecto.Changeset.put_assoc(rider_changeset, :tags, [rapid_response_tag])
```

Comment:
There's a whole world of put_assoc vs cast_assoc that I find super confusing. I don't think you should explain it all here but worth linking to the docs / mentioning it